storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -62,123 +70,144 @@ public void prepare(Map<String, Object> conf) {
+                        List<TopologyDetails> reversedList = ImmutableList.copyOf(orderedTopologies).reverse();
                         try {
-                            //need to re prepare since scheduling state might have been restored
-                            madeSpace = evictionStrategy.makeSpaceForTopo(td, workingState, userMap);
+                            boolean evictedSomething = false;
+                            LOG.debug("attempting to make space for topo {} from user {}", td.getName(), td.getTopologySubmitter());
+                            int tdIndex = reversedList.indexOf(td);
+                            double cpuNeeded = td.getTotalRequestedCpu();
+                            double memoryNeeded = td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap();
+                            SchedulerAssignment assignment = cluster.getAssignmentById(td.getId());
+                            if (assignment != null) {
+                                cpuNeeded -= getCpuUsed(assignment);
+                                memoryNeeded -= getMemoryUsed(assignment);
+                            }
+                            cluster.getTopologyResourcesMap();
+                            for (int index = 0; index < tdIndex; index++) {
+                                TopologyDetails topologyEvict = reversedList.get(index);
+                                SchedulerAssignment evictAssignemnt = workingState.getAssignmentById(topologyEvict.getId());
+                                if (evictAssignemnt != null && !evictAssignemnt.getSlots().isEmpty()) {
+                                    Collection<WorkerSlot> workersToEvict = workingState.getUsedSlotsByTopologyId(topologyEvict.getId());
+
+                                    LOG.debug("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict,
+                                        topologyEvict.getTopologySubmitter());
+                                    cpuNeeded -= getCpuUsed(evictAssignemnt);
+                                    memoryNeeded -= getMemoryUsed(evictAssignemnt);
+                                    evictedSomething = true;
+                                    nodes.freeSlots(workersToEvict);
+                                    if (cpuNeeded <= 0 && memoryNeeded <= 0) {
+                                        //We evicted enough topologies to have a hope of scheduling, so try it now, and don't evict more
+                                        // than is needed
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (!evictedSomething) {
+                                markFailedTopology(topologySubmitter, cluster, td,
+                                    "Not enough resources to schedule - " + result.getErrorMessage());
```

Comment:
Should we change this to give them an idea of what resource was missing?  CPU/Memory etc.